### DISCUSSION TITLE
Table relationships without Postgres owner rights

### DISCUSSION THREAD

**Discussion by joshuajung** [directus/directus#7244] [Archive]
Hi there,

I have been exploring Directus (`v9.0.0-rc.88`) over the last few days and am pretty delighted about this product! However, I have encountered the following issue, which I would like to run by you more experienced folks:

I have an existing backend with its legacy Postgres table structure. I would like to use Directus as a back office UI, but would like to stop it (and myself) from inadvertently modifying our database table structure.   Therefore, I have provided Directus with a Postgres user without owner rights for my content tables. This should allow it to work with records as it pleases, but stops it from adding columns, indexes, etc. The user has full ownership rights for the `directus_*` tables.

First question: is this a stupid idea? Everything seems to work for the moment.   

Well, except one thing: When I try to set display settings in Directus for a data model field that has a relationship to another table, I get this error:

```
[INTERNAL_SERVER_ERROR] alter table "apartments" drop constraint "fk_apartments_buildings_building_id" - must be owner of table apartments
```

So it seems like Directus tries to modify my table constraints, even though this was never my intention – I just want to tell it how to display related records, and I'm not even touching the _Relationship_ tab.

Is this working as intended? Is there a way to make relationships look good in Directus without full Postgres owner rights for the respective tables?

Any input would be greatly appreciated. Thank you!

**Comment by rijkvanzanten** [comment]
> First question: is this a stupid idea? Everything seems to work for the moment.

Not a stupid idea at all! That being said, Directus does expect being able to modify the table schema when you make alterations to the schema through settings (regardless of whether or not it touches the database column), as it will also try to clean up any potential corruptions on those updates. 

All of this in mind, I think it's probably best to treat this as a feature request for a new "don't modify the schema" flag in the environment variables, that effectively disables schema manipulation system wide, instead of trying to rely on the database user (as those permissions could be very different on a case to case /db to db vendor basis 🤔

**Reply by joshuajung** [reply]
Yes, that sounds like a great feature idea to me! I'd be totally happy to use Directus with half of the features disabled (those relying on schema modifications outside the `directus_*` tables), because the UI is so enjoyable already. 

Thanks for your quick reply!

**Reply by rijkvanzanten** [reply]
It also ties into https://github.com/directus/directus/discussions/3891 fairly well (TL;DR: adding a way to setup the schema through a schema-file instead of through the app/api) I think 🤔 

(cc @benhaynes)

**Reply by benhaynes** [reply]
I'm down to have a flag to lock schema edits... and anything to round out the schema migrations is a great idea. 👍

**Comment by joshuajung** [comment]
Just in case somebody is trying to do the same, here is my workaround (probably not recommended, but it worked fine for me for now):

I patched a local instance of Directus by commenting out these lines in `node_modules/directus/dist/services/relations.js` (starting in line 197 for rc.92):
```js
            if (existingRelation.related_collection) {
                await trx.schema.alterTable(collection, async (table) => {
                    var _a;
                    let constraintName = (0, get_default_index_name_1.getDefaultIndexName)('foreign', collection, field);
                    // If the FK already exists in the DB, drop it first
                    if (existingRelation === null || existingRelation === void 0 ? void 0 : existingRelation.schema) {
                        constraintName = existingRelation.schema.constraint_name || constraintName;
                        table.dropForeign(field, constraintName);
                    }
                    this.alterType(table, relation);
                    const builder = table
                        .foreign(field, constraintName || undefined)
                        .references(`${existingRelation.related_collection}.${this.schema.collections[existingRelation.related_collection].primary}`);
                    if ((_a = relation.schema) === null || _a === void 0 ? void 0 : _a.on_delete) {
                        builder.onDelete(relation.schema.on_delete);
                    }
                });
            }
```

Now I can format the relationships as desired without any risk of ruining the original schema. 🎉
